File: osint_nexus/core/detectors/extensions.py
from collections.abc import Mapping
from beartype import beartype

from osint_nexus.core.detectors.base import FingerprintStrategy
from osint_nexus.core.type_defs import JSONValue


class ExtensionFingerprintStrategy(FingerprintStrategy[Mapping[str, JSONValue], Mapping[str, JSONValue]]):
    """Strategy for Browser Extension fingerprinting."""

    name: str = "extension_load"

    @beartype
    def extract(self, data: Mapping[str, JSONValue]) -> Mapping[str, JSONValue]:
        # Expecting data: {"detected_extensions": list[str]}
        extensions_raw = data.get("detected_extensions")
        extensions = extensions_raw if isinstance(extensions_raw, list) else []

        # Heuristic: analyze extension load order or set
        has_adblocker = any(
            isinstance(e, str) and ("adblock" in e.lower() or "ublock" in e.lower()) for e in extensions
        )

        fingerprint: Mapping[str, JSONValue] = {
            "extension_count": len(extensions),
            "has_adblocker": has_adblocker,
        }

        confidence = 0.8 if len(extensions) > 0 else 0.2

        return {
            "name": self.name,
            "data": fingerprint,
            "confidence": confidence,
        }

File: internal/engine/strategies/extensions.go
package strategies

import (
	"context"
	"strings"

	"github.com/FJ-cyberzilla/osint-nexus/internal/types"
)

// ExtensionPayload implements types.FingerprintPayload for extension fingerprinting input.
type ExtensionPayload struct {
	DetectedExtensions []string `json:"detected_extensions"`
}

// PayloadType returns the payload type identifier.
func (p ExtensionPayload) PayloadType() string { return "extension_input" }

// ExtensionOutputPayload implements types.FingerprintPayload for extension fingerprinting output.
type ExtensionOutputPayload struct {
	ExtensionCount int  `json:"extension_count"`
	HasAdblocker   bool `json:"has_adblocker"`
}

// PayloadType returns the payload type identifier.
func (p ExtensionOutputPayload) PayloadType() string { return "extension_output" }

// ExtensionFingerprintStrategy identifies browser extension load signatures.
type ExtensionFingerprintStrategy struct{}

// NewExtensionFingerprintStrategy initializes a new ExtensionFingerprintStrategy.
func NewExtensionFingerprintStrategy() *ExtensionFingerprintStrategy {
	return &ExtensionFingerprintStrategy{}
}

// Name returns the strategy identifier.
func (s *ExtensionFingerprintStrategy) Name() string {
	return "extension_load"
}

// Extract analyzes detected extensions.
func (s *ExtensionFingerprintStrategy) Extract(ctx context.Context, data types.FingerprintData) (types.FingerprintResult, error) {
	payload, ok := data.Payload.(ExtensionPayload)
	if !ok {
		return types.FingerprintResult{
			Name: s.Name(),
			Data: types.FingerprintData{
				Type:    "extension_detection",
				Payload: ExtensionOutputPayload{ExtensionCount: 0, HasAdblocker: false},
			},
			Confidence: 0.2,
		}, nil
	}

	hasAdblocker := false
	for _, e := range payload.DetectedExtensions {
		lowerE := strings.ToLower(e)
		if strings.Contains(lowerE, "adblock") || strings.Contains(lowerE, "ublock") {
			hasAdblocker = true
			break
		}
	}

	confidence := 0.2
	if len(payload.DetectedExtensions) > 0 {
		confidence = 0.8
	}

	return types.FingerprintResult{
		Name: s.Name(),
		Data: types.FingerprintData{
			Type: "extension_detection",
			Payload: ExtensionOutputPayload{
				ExtensionCount: len(payload.DetectedExtensions),
				HasAdblocker:   hasAdblocker,
			},
		},
		Confidence: confidence,
	}, nil
}
